[mmp] Add stripping of 32-bit dylibs to work with new App Store restrictions

chamons opened the pull request:
- https://github.com/xamarin/xamarin-macios/issues/3367
- Required some test work that might be more obvious by themselves. We can squash when landing.

monojenkins commented:
Build failure

spouliot commented on tools/mmp/driver.cs:
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
+				string archToRemove = arch == "i386" ? "x86_64" : "i386";
+				int ret = XcodeRun ("lipo", $"{dest} -remove {archToRemove} -output {dest}");
+				if (ret != 0)
+					throw new MonoMacException (5310, true, "lipo failed with an error code '{0}'. Check build log for details.", ret);
```
is this a new error? if so it should be documented

chamons replied:
Yes, I forgot.

spouliot commented on tools/mmp/driver.cs:
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
+				if (ret != 0)
+					throw new MonoMacException (5310, true, "lipo failed with an error code '{0}'. Check build log for details.", ret);
+				if (name != "MonoPosixHelper")
+					ErrorHelper.Warning (1501, $"{name} was stripped of architecture {archToRemove} to comply with App Store restrictions. This could break exisiting codesigning signatures. Consider stripping the library with lipo or disabling with --native-strip=skip");
```
same

spouliot commented on tools/mmp/driver.cs:
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
-
+			
 			bool isStaticLib = real_src.EndsWith (".a", StringComparison.Ordinal);
+			bool isDynamicLib = real_src.EndsWith (".dylib", StringComparison.Ordinal);
```
what happens for frameworks ?

chamons replied:
Nothing, but something very likely should. Let me get someone in to test this case in submission to Apple.

rolfbjarne commented on tests/mmptest/src/CodeStrippingTests.cs:
```diff
@@ -0,0 +1,60 @@
+
+				 test.Release = true;
+				 buildOutput = TI.TestUnifiedExecutable (test).BuildOutput;
+				 Assert.AreEqual (releaseStrips, didStrip (buildOutput), "Release lipo usage did not match expectations");
```
Maybe also test that the binary only has the desired architectures? We already have code to get the architectures of a binary file: https://github.com/xamarin/xamarin-macios/blob/bec116c273ab119c648522a28c3f650881ed4d9e/tests/mtouch/MTouch.cs#L3996

rolfbjarne commented on tests/common/mac/ProjectTestHelpers.cs:
```diff
@@ -145,11 +152,15 @@ public static string BuildProject (string csprojTarget, bool isUnified, bool dia
-				if (!string.IsNullOrEmpty (configuration))
-					buildArgs.Append ($" /property:Configuration={configuration} ");
-			} else
+			}
```
Style: `else` goes on the same line as the closing brace.

rolfbjarne commented on tools/mmp/driver.cs:
```diff
@@ -1685,8 +1712,19 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
 				CopyFileAndRemoveReadOnly (real_src, dest);
 			}
-
+			
```
Whitespace noise.

monojenkins commented:
Build failure

chamons commented:
I believe I've handled the current code review issues and added framework stripping.

I did not build after merging master, and a mono bump took too long before end of day. I'll check back tomorrow. Please consider re-reviewing as a lot has changed.

monojenkins commented:
Build failure

monojenkins commented:
Build failure

spouliot commented on tools/mmp/driver.cs:
```diff
@@ -366,6 +374,22 @@ static void Main2 (string [] args)
+					default:
+						ErrorHelper.Error (26, $"Could not parse the command line argument '-native-strip:{v}'");
+						break;
+					}
```
I wonder if that should be added in the new optimizations ? c.c. @rolfbjarne 
It behaves just like them, i.e. a sane *default* (computed based on the project settings) and overrides for *enabling* and *disabling* it.
In any case it needs to be documented outside the errors code it introduce.

chamons commented on tools/mmp/driver.cs:
```diff
@@ -1720,6 +1755,17 @@ static void ProcessNativeLibrary (HashSet<string> processed, string library, Lis
+			if (ret != 0)
+				throw new MonoMacException (5311, true, "lipo failed with an error code '{0}'. Check build log for details.", ret);
+			if (name != "MonoPosixHelper")
+				ErrorHelper.Warning (2108, $"{name} was stripped of architecture {archToRemove} to comply with App Store restrictions. This could break exisiting codesigning signatures. Consider stripping the library with lipo or disabling with --native-strip=skip");
```
The "obvious" thing to do here is to first write a test that says "if I'm 64-bit and i have a 64-bit library there should be no warning".

And yes, it looks like that test would fail right now. :(

chamons commented:
I've moved to --optimize=-trim-architectures instead of a new option and fixed the location of common string utils.

I'm going to work on fixing the likely every present warning w\ native libraries.

@rolfbjarne please review my optimization changes, first time I've added one.

monojenkins commented:
Build failure

monojenkins commented:
Build failure

spouliot commented on tools/common/Driver.cs:
```diff
@@ -142,7 +142,12 @@ static void AddSharedOptions (Application app, Mono.Options.OptionSet options)
-					"    register-protocols: Remove unneeded metadata for protocol support. Makes the app smaller and reduces memory requirements.\n",
+					"    register-protocols: Remove unneeded metadata for protocol support. Makes the app smaller and reduces memory requirements.\n" +
+#if !MONOTOUCH
+					"    trim-architectures: Remove unneeded architectures from bundled native libraries. Makes the app smaller and necessary for macOS App Store submissions.\n",
```
**required** ? more than necessary ?

chamons commented on docs/website/mmp-errors.md:
```diff
@@ -248,6 +250,15 @@ To remove the warning either remove the optimization argument to mmp, or pass
+
+The App Store now rejects applications which contain libraries and frameworks which contain 32-bit variants. The library was stripped of unused archtectures when copied into the final application bundle.
+
+This is in general safe, and will reduce application size as an added benefit, but may break existing codesigning signatures if it was already signed.
```
We do, but if there was some pre-existing signature on the frameowork we're copying we'd be breaking that and resigning over top.

Maybe it isn't worth mentioning that corner case. In 99% of use cases we're going to fix things when signing at the end.

rolfbjarne replied:
My point is that if we're always signing dylibs and frameworks, stripping unused architectures from them won't change anything (since we always overwrite any existing signature anyway).

rolfbjarne replied:
Ah, I see, it's not required to sign the .app like it is for XI, in which case this does make sense. So nvm.

monojenkins commented:
Build failure

chamons commented:
Ok, I added ThirdPartyLibrary_WithCorrectBitness_ShouldNotStripOrWarn test and fixed the others so they would fail without my fix.

I could make the logic more complicated in mmp but bailing on < 2 arch in a lib works and seemed simpler. Thoughts?

chamons commented:
Some test failures are legit and need looking at.

chamons commented:
I can not reproduce them currently with ```make run-mac-unified-mmptest``` or mmptest.csproj. Trying another build in parallel.

chamons commented:
build

monojenkins commented:
Build failure

monojenkins commented:
Build failure

chamons commented:
I can reproduce them when run from the command line but not from the IDE, which is strange. In any case, looking.

monojenkins commented:
Build failure

chamons commented:
This PR has started to turn into a monster, just like every other time I touch mmp...

A few more tests are failing locally after the last set of pushes. Looking now.

chamons commented:
Looks clean enough locally now. 🤞

monojenkins commented:
Build failure

monojenkins commented:
Build failure

chamons commented:
Looks like some tests are comparing strings that changed when I added a message. Fixing.

monojenkins commented:
Build failure

monojenkins commented:
Build failure

chamons commented:
Some known test failures:

https://github.com/xamarin/maccore/issues/640
https://github.com/xamarin/maccore/issues/656
https://github.com/xamarin/maccore/issues/629

Looking into protocol failures, feel unrelated.

chamons commented:
https://github.com/xamarin/maccore/issues/655 for the protocol failures.
